File: 2020/09.rb
PREAMBLE_SIZE = 25
nums = File.readlines('09_input.txt').map(&:to_i)

# Part 1
first_invalid_num = PREAMBLE_SIZE.upto(nums.length - 1).each do |i|
  n = nums[i]
  preamble = nums[i - PREAMBLE_SIZE ... i]
  break n if !preamble.product(preamble).map(&:sum).include?(n)
end
# => 552655238


# Part 2
0.upto(nums.length - 1) do |i|
  span = [nums[i], nums[i + 1]]
  next_in_span = 2

  loop do
    break if span.sum >= first_invalid_num
    span << nums[i + next_in_span]
    next_in_span += 1
  end

  break span.min + span.max if span.sum == first_invalid_num
end
# => 70672245

File: 2020/09.ml
#use "helpers.ml";;

let self_product arr =
  let scalar_product x acc =
    Array.fold_right (fun y acc2 -> Array.append [|[|x; y|]|] acc2) arr acc in
  Array.fold_right scalar_product arr [||];;

let range a b = S.range a b |> List.of_seq;;

let preamble_len = 25;;
let numbers = read_lines "09_input.txt"
              |> List.map int_of_string
              |> Array.of_list;;
let len = Array.length numbers;;

(* Part 1 *)
let check_num i =
  let n = numbers.(i)
  and preamble = Array.sub numbers (i - preamble_len) preamble_len in
  if not (Array.map A.sum (self_product preamble) |> Array.mem n) then
    Some n
  else
    None;;

List.find_map check_num (range preamble_len len);;
(* - : int option = Some 552655238 *)


(* Part 2 *)
let first_invalid_number = 552655238;;

let check_sum i =
  let check_span j =
    let span = Array.sub numbers i (j + 2) in
    let n = span |> A.sum in
    if n >= first_invalid_number then
      Some (n, span)
    else
      None in
  let result = List.find_map check_span (range 0 (len - i - 2)) in
  match result with
  | None -> None
  | Some (n, span) ->
    if n = first_invalid_number then
      Some ((A.min span) + (A.max span))
    else
      None;;

List.find_map check_sum (range 0 len);;
(* - : int option = Some 70672245 *)
